distributed/core.py

```diff
@@ -109,6 +109,144 @@ def _expects_comm(func: Callable) -> bool:
     return False
 
 
+class _LoopBoundMixin:
+    """Backport of the private asyncio.mixins._LoopBoundMixin from 3.11"""
+
+    _global_lock = threading.Lock()
+
+    _loop = None
+
+    def _get_loop(self):
+        loop = asyncio.get_running_loop()
+
+        if self._loop is None:
+            with self._global_lock:
+                if self._loop is None:
+                    self._loop = loop
+        if loop is not self._loop:
+            raise RuntimeError(f"{self!r} is bound to a different event loop")
+        return loop
+
+
+class AsyncTaskGroup(_LoopBoundMixin):
+    """Collection tracking all currently running asynchronous tasks within a group"""
+
+    #: If True, the group is closed and does not allow adding new tasks.
+    closed: bool
+
+    def __init__(self) -> None:
+        self.closed = False
+        self._ongoing_tasks: set[asyncio.Task] = set()
+
+    def schedule(self, coro: Coroutine) -> asyncio.Task | None:
+        """Schedules a coroutine object to be executed as an `asyncio.Task`.
+
+        Parameters
+        ----------
+        coro
+            Coroutine object to schedule.
+
+        Returns
+        -------
+            The scheduled Task object, or None if the group is closed.
+        """
+        if self.closed:
+            coro.close()
+            return None
```

Comment:
can we do something like this here?

```python
class AsyncTaskGroupClosedError(RuntimeError):
    pass
```

```suggestion
            raise AsyncTaskGroupClosedError(...)
```

then we can solve the unawaited coroutine in handle_comm with:
```python
                            try:
                                task = self._ongoing_comm_handlers.schedule(lambda: result)
                            except AsyncTaskGroupClosedError:
                                result.close()  # TODO don't call coroutinefunctions we can't await
                            result = await task
```

?